### PR TITLE
Internal changes to allow for new features

### PR DESCRIPTION
**Pull request by thomasbird**
This MR changes the insides of scrubadub a little to allow for several new features (while aiming to keep backwards compatibility with the old version).

* Introduced the concept of a `PostProcessor`. This will allow more complex groupings of `Filth`s and new types of tokenization.
* Added ability to easily evaluate a `Detector`'s performance.
* Added functionality to clean multiple documents at once.
* The the name of the detector has been separated from the type of filth found. This means multiple instances of the same detector (configured differently) can be in the same `Scrubber` instance and one `Detector` can return multiple types of `Filth`.
* A default set of `Detector`s are loaded instead of all `Detector`s. This is particularly useful for optional detectors with complex dependencies.

Further detail can be found in [the changelog](https://github.com/LeapBeyond/scrubadub/blob/feature/api_changes/docs/changelog.rst) 

This is a big changeset, so I welcome any feedback. :) @aCampello

### CODE REVIEW

**Comment by aCampello** [comment]
I really like the new API. The logic `Scrubber -> [Detector, Post-Processor](List_of_filths)` makes a lot of sense for me. I wasn't able to read the new post-processors that carefully. A couple of minor comments

- There is one `mypy` test that fails locally (see comment in the code). Weirdly it doesn't fail on your Travis CI...
- A couple of complaints by my linter (ignoring line-length, which I think you are not using). Not sure if you want to address those separately after merging the PR, but here it goes
```
scrubadub/filth/email.py:1:1: F401 're' imported but unused
scrubadub/filth/skype.py:1:1: F401 're' imported but unused
scrubadub/detectors/ssn.py:1:1: F401 're' imported but unused
scrubadub/detectors/known.py:3:1: F401 'typing.Union' imported but unused
scrubadub/detectors/known.py:3:1: F401 'typing.Dict' imported but unused
scrubadub/detectors/base.py:4:1: F401 '..exceptions' imported but unused
scrubadub/post_processors/base.py:1:1: F401 'typing.Type' imported but unused
scrubadub/post_processors/base.py:1:1: F401 'typing.List' imported but unused
scrubadub/post_processors/text_replacers/hash.py:5:1: F401 'typing.ClassVar' imported but unused
scrubadub/post_processors/text_replacers/hash.py:5:1: F401 'typing.List' imported but unused
scrubadub/post_processors/text_replacers/numeric.py:1:1: F401 'os' imported but unused
scrubadub/post_processors/text_replacers/numeric.py:2:1: F401 'hashlib' imported but unused
scrubadub/post_processors/text_replacers/numeric.py:4:1: F401 'typing.Optional' imported but unused
scrubadub/post_processors/text_replacers/numeric.py:4:1: F401 'typing.Type' imported but unused
scrubadub/post_processors/text_replacers/numeric.py:4:1: F401 'typing.List' imported but unused
scrubadub/post_processors/text_replacers/filth_type.py:1:1: F401 'os' imported but unused
scrubadub/post_processors/text_replacers/filth_type.py:2:1: F401 'hashlib' imported but unused
```

- Some of the detectors' regular expressions are defined in `./filth` (e.g. credential) and some others in `./detectors` (e.g. email). Is there any reason for it? As far as I am concerned, they should all be detectors...


But as far as I am concerned we can merge the PR. Then I will think about picking up some of the issues on top of the new API...

**Comment by thomasbird** [comment]
Hey @aCampello,

Thanks for having a look through! 

- I was unable to find a comment on the code about mypy, where should I be looking?
- I'll fix up those imports quickly, thanks for pointing these out
- On the detector regex's they used to be part of the `Filth` class and I was moving them to the `Detector` class, but I hadn't moved them all. I'll move the rest now. 

I'll push a commit or two to fix these shortly.

Thanks,
Thomas

**Comment by aCampello** [comment]
> * I was unable to find a comment on the code about mypy, where should I be looking?

Apologies, I forgot to click "submit"! Hopefully you can see the comments now.

File: scrubadub/scrubbers.py [hunk 1]
```diff
@@ -74,26 +233,94 @@ def iter_filth(self, text):
...
-        filth = all_filths[0]
-        for next_filth in all_filths[1:]:
-            if filth.end < next_filth.beg:
-                yield filth
-                filth = next_filth
-            else:
-                filth = filth.merge(next_filth)
-        yield filth
+
+        document_names = sorted({f.document_name for f in filth_list})
```
[the remark this answers is not in the document]
Comment (thomasbird) [reply scrubadub/scrubbers.py:314]: What version of mypy and python do you have? No idea what's going on there. I'll rewrite the line.

[the remark this answers is not in the document]
Comment (aCampello) [reply scrubadub/scrubbers.py:314]: I found the problem - I was running with `mypy-0.790`. When I downgraded to `0.782` (which is the same that your Travis-ci) it passed! Up to you, but I think the new version  is correct in saying that that line has an error...

[the remark this answers is not in the document]
Comment (thomasbird) [reply scrubadub/scrubbers.py]: Aha, after an upgrade, I too see the issue. Thanks for spotting this! The fix is now committed.

**Pull request merged**